drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
+///   - Entirely in flash
+///   - Header block in RAM with the remainder unavailable
+///   - Header block in RAM with the remainder in flash.
+///   - Entire image in RAM (cached Stage0)
+/// Calls to methods use image relative addresses.
+/// ImageAccess::Flash{_, span} gives the physical flash addresses
+/// being accessed.
+///
+/// Any address here is the "storage" address which for stage0next
+/// and any image held temporarily in RAM is different than the
+/// execution address.
+pub struct ImageAccess<'a> {
+    accessor: Accessor<'a>,
+}
+
+impl ImageAccess<'_> {
+    pub fn new_flash<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'a> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Flash { flash, span },
+        }
+    }
+
+    pub fn new_ram(
+        buffer: &[u8],
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'_> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Ram { buffer, span },
+        }
+    }
+
+    pub fn _new_hybrid<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        buffer: &'a [u8],
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'a> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::_Hybrid {
+                flash,
+                buffer,
+                span,
+            },
+        }
+    }
+
+    fn exec(&self) -> &Range<u32> {
+        self.accessor.exec()
+    }
+
+    /// True if the u32 at offset is contained within the image.
+    pub fn is_addressable(&self, offset: u32) -> bool {
+        let len = match &self.accessor {
+            Accessor::Flash { flash: _, span }
+            | Accessor::_Hybrid {
+                buffer: _,
+                flash: _,
+                span,
+            }
+            | Accessor::Ram { buffer: _, span } => {
+                span.store.end.saturating_sub(span.store.start)
+            }
+        };
+        offset < len && offset.saturating_add(U32_SIZE) <= len
+    }
+
+    // Fetch a u32 from an image.
+    pub fn read_word(&self, offset: u32) -> Result<u32, UpdateError> {
+        if !self.is_addressable(offset) {
+            return Err(UpdateError::OutOfBounds);
+        }
+        match &self.accessor {
+            Accessor::Flash { flash, span } => {
+                let addr = span.store.start.saturating_add(offset);
+                let mut word = 0u32;
+                indirect_flash_read(flash, addr, word.as_bytes_mut())?;
+                Ok(word)
+            }
+            Accessor::Ram { buffer, span: _ } => {
+                match buffer
+                    .get(offset as usize..(offset as usize + SIZEOF_U32))
+                    .and_then(u32::read_from)
+                {
+                    Some(word) => Ok(word),
```

Comment:
FWIW this is equivalent to `.ok_or(UpdateError::OutOfBounds)?`

Reply:
done